src/main/scala/com/microsoft/hyperspace/index/rules/IndexFilter.scala

```diff
@@ -74,3 +80,151 @@ trait RankerIndexFilter extends IndexFilter with ActiveSparkSession {
+      spark: SparkSession,
+      plan: LogicalPlan,
+      indexes: Seq[IndexLogEntry]): Unit = {
+    assert(HyperspaceConf.hybridScanEnabled(spark))
+    val curConfigs = Seq(
+      HyperspaceConf.hybridScanAppendedRatioThreshold(spark).toString,
+      HyperspaceConf.hybridScanDeletedRatioThreshold(spark).toString)
+
+    indexes.foreach { index =>
+      val taggedConfigs = index.getTagValue(plan, HYBRIDSCAN_RELATED_CONFIGS)
+      if (taggedConfigs.isEmpty || !taggedConfigs.get.equals(curConfigs)) {
+        // Need to reset cached tags as these config changes can change the result.
+        index.unsetTagValue(plan, IS_HYBRIDSCAN_CANDIDATE)
+        index.setTagValue(plan, HYBRIDSCAN_RELATED_CONFIGS, curConfigs)
+      }
+    }
+  }
+
+  private def getHybridScanCandidate(
+      relation: FileBasedRelation,
+      index: IndexLogEntry,
+      hybridScanDeleteEnabled: Boolean): Option[IndexLogEntry] = {
+    // TODO: As in [[PlanSignatureProvider]], Source plan signature comparison is required to
+    //  support arbitrary source plans at index creation.
+    //  See https://github.com/microsoft/hyperspace/issues/158
+
+    val entry = relation.closestIndex(index)
+
+    val isHybridScanCandidate =
+      entry.withCachedTag(relation.plan, IndexLogEntryTags.IS_HYBRIDSCAN_CANDIDATE) {
+        // Find the number of common files between the source relation and index source files.
+        // The total size of common files are collected and tagged for candidate.
+        val (commonCnt, commonBytes) = relation.allFileInfos.foldLeft(0L, 0L) { (res, f) =>
+          if (entry.sourceFileInfoSet.contains(f)) {
+            (res._1 + 1, res._2 + f.size) // count, total bytes
+          } else {
+            res
+          }
+        }
+
+        val appendedBytesRatio = 1 - commonBytes / relation.allFileSizeInBytes.toFloat
+        val deletedBytesRatio = 1 - commonBytes / entry.sourceFilesSizeInBytes.toFloat
+
+        val deletedCnt = entry.sourceFileInfoSet.size - commonCnt
+        val isAppendAndDeleteCandidate = hybridScanDeleteEnabled && entry.hasLineageColumn &&
+          commonCnt > 0 &&
+          appendedBytesRatio < HyperspaceConf.hybridScanAppendedRatioThreshold(spark) &&
+          deletedBytesRatio < HyperspaceConf.hybridScanDeletedRatioThreshold(spark)
+
+        // For append-only Hybrid Scan, deleted files are not allowed.
+        lazy val isAppendOnlyCandidate = deletedCnt == 0 && commonCnt > 0 &&
+          appendedBytesRatio < HyperspaceConf.hybridScanAppendedRatioThreshold(spark)
+
+        val isCandidate = isAppendAndDeleteCandidate || isAppendOnlyCandidate
+        if (isCandidate) {
+          entry.setTagValue(
+            relation.plan,
+            IndexLogEntryTags.COMMON_SOURCE_SIZE_IN_BYTES,
+            commonBytes)
+
+          // If there is no change in source dataset, the index will be applied by
+          // RuleUtils.transformPlanToUseIndexOnlyScan.
+          entry.setTagValue(
```

Comment:
Let's clearly document in the public API if we are modifying the entry. Also, should this be done at the `SourcePlanIndexFilter` level or do we need a different abstraction?

Reply:
These tags are for caching result, so I guess it's hard to do this in somewhere else.